### PR TITLE
Alibaba Cloud BCS backend and OSS storage nio interface

### PR DESCRIPTION
**Pull request by crisish**
1. BCS supported backend.
2. OSS storage nio interface.

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 1]
```diff
@@ -134,7 +134,7 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
     WomSingleFile(workflowPaths.buildPath(womFile.value).pathAsString)
 
   /** @see [[Command.instantiate]] */
-  final lazy val commandLineValueMapper: WomValue => WomValue = {
+  lazy val commandLineValueMapper: WomValue => WomValue = {
```
Comment (crisish) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: I remove final keyword because I can not find a more elegant way to map some oss path string in command line, for example: 
task ps {
  String out

  command {
    ps > ${out}
  }

  output {
    File outFile = "${out}"
  }
}

workflow three_step {
  call ps {
    input: out = "oss://bcs-dir/bcs-file"
  }
}

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssStorageFileAttributesView.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import scala.util.Try
+
+case class OssStorageFileAttributesView(ossClient: OSSClient, path: OssStoragePath) extends BasicFileAttributeView {
+  override def name(): String = OssStorageFileSystem.URI_SCHEMA
+
+  override def readAttributes(): OssStorageFileAttributes = {
+    val ossPath = OssStoragePath.checkPath(path)
+
+    if (ossPath.seemsLikeDirectory) {
+      return new OssStorageDirectoryAttributes(path)
```
Comment (danbills) [inline filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssStorageFileAttributesView.scala]: no need for `return` keyword

File: project/Dependencies.scala [hunk 3]
```diff
@@ -206,6 +206,19 @@ object Dependencies {
     "org.apache.httpcomponents" % "httpclient" % apacheHttpClientV
   )
 
+  private val aliyunOssDependencies = List(
+    "com.aliyun.oss" % "aliyun-sdk-oss" % "2.8.3"
```
Comment (danbills) [inline project/Dependencies.scala]: could you please set the version as a `val` as is our convention?

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        .getOrElse(List.empty)
+    }
+
+    // val womFileOutputs = call.task.findOutputFiles(jobDescriptor.fullyQualifiedInputs, PureStandardLibraryFunctions)
+
+    jobDescriptor.taskCall.callable.outputs.flatMap(evaluateFiles)
+  }
+
+  private[bcs] def isOutputOssFileString(s: String): Boolean = {
+    println(callRawOutputFiles)
```
Comment (danbills) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: Please either log or remove.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+  private lazy val inputMounts: ArrayBuffer[BcsMount] = ArrayBuffer.empty
+
+  private[bcs] def ossPathToMount(ossPath: OssPath): BcsInputMount = {
+    val tmp = DefaultPathBuilder.get("/" + ossPath.pathWithoutScheme)
+    val dir = tmp.getParent
+    val local = BcsJobPaths.BcsTempInputDirectory.resolve(dir.pathAsString.md5SumShort).resolve(tmp.getFileName)
+    val ret = BcsInputMount(ossPath, local, writeSupport = false)
+    if (!inputMounts.exists(mount => mount.src == ossPath && mount.dest == local)) {
+      inputMounts += ret
+      println(inputMounts)
```
Comment (danbills) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: see below re: `println` and please apply this principle throughout.

File: engine/src/main/scala/cromwell/engine/io/IoActor.scala [hunk 6]
```diff
@@ -57,19 +61,24 @@ final class IoActor(queueSize: Int,
     val input = builder.add(Flow[IoCommandContext[_]])
     
     // Partitions requests between gcs batch, and single nio requests
-    val batchPartitioner = builder.add(Partition[IoCommandContext[_]](2, {
-      case _: GcsBatchCommandContext[_, _] => 0
-      case _ => 1
+    val batchPartitioner = builder.add(Partition[IoCommandContext[_]](3, {
+      case _: OssCommandContext[_] => 0
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/io/IoActor.scala]: Since you implemented the Nio interface for the OSS filesystem, I don't think any of those modifications are necessary.
The `GcsXXX` classes (except `GcsPath`) are only optimization for GCS. The `DefaultIoCommand`s will use the nio interface to perform the corresponding operations.

File: engine/src/main/scala/cromwell/engine/io/oss/OssFlow.scala [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import cromwell.engine.io.IoActor
+import cromwell.engine.io.IoActor.{IoResult, OssCommandContext}
+import cromwell.filesystems.oss._
+import java.nio.file.{Files, Paths}
+
+import better.files.File.OpenOptions
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class OssFlow(parallelism: Int, scheduler: Scheduler, nbAttempts: Int = IoActor.MaxAttemptsNumber)(implicit ec: ExecutionContext, actorSystem: ActorSystem) {
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/io/oss/OssFlow.scala]: This should not be necessary either if you stick to the Nio implementation.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+  private[bcs] def dataDisk: Option[DataDisk] = runtime.dataDisk map { disk =>
+    val dataDisk = new DataDisk
+    dataDisk.setType(disk.diskType)
+    dataDisk.setSize(disk.sizeInGB)
+    dataDisk.setMountPoint(disk.mountPoint)
+    dataDisk
+  }
+
+  // XXX: maybe more elegant way to reduce two options?
```
Comment (danbills) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala]: perhaps
```scala
systemDisk.foreach(lazyDisks.setSystemDisk)
dataDisk.foreach(lazyDisks.setDataDisk)
```

File: filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala [hunk 9]
```diff
@@ -34,4 +42,14 @@ trait GcsBatchCommandBuilder extends DefaultIoCommandBuilder {
...
   }
+
+  override def writeCommand(path: Path, content: String, options: OpenOptions) = path match {
+    case ossPath: OssPath => OssIoWriteCommand(ossPath, content, options)
+    case _ => super.writeCommand(path, content, options)
+  }
+
+  override def contentAsStringCommand(path: Path) = path match {
+    case ossPath: OssPath => OssIoContentAsStringCommand(ossPath)
+    case _ => super.contentAsStringCommand(path)
```
Comment (Horneth) [inline filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala]: Same here, now that you implemented the nio interface you stick to `DefaultIoCommand`s and it will all work

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/OssIoCommand.scala [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.nio.file.OpenOption
+
+import better.files.File.OpenOptions
+import cromwell.core.io._
+
+case object UploadFileOption extends OpenOption
+case object UploadStringOption extends OpenOption
+
+trait OssIoCommand[T] extends IoCommand[T]
```
Comment (Horneth) [inline filesystems/oss/src/main/scala/cromwell/filesystems/oss/OssIoCommand.scala]: No need for this either (see comments above)

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/OssPathBuilder.scala [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  sealed trait InvalidOssPath extends OssPathValidation {
+    def pathString: String
+    def errorMessage: String
+  }
+
+  final case class InvalidScheme(pathString: String) extends InvalidOssPath {
+    def errorMessage = s"OSS URIs must have 'oss' scheme: $pathString"
+  }
+
+  final case class InvalidFullGcsPath(pathString: String) extends InvalidOssPath {
```
Comment (Horneth) [inline filesystems/oss/src/main/scala/cromwell/filesystems/oss/OssPathBuilder.scala]: `Gcs` -> `Oss`

File: filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala [hunk 12]
```diff
@@ -1,31 +1,39 @@
 package cromwell.filesystems.gcs.batch
 
 import cromwell.core.io._
+import cromwell.core.path.BetterFileMethods.OpenOptions
 import cromwell.core.path.Path
 import cromwell.filesystems.gcs.GcsPath
+import cromwell.filesystems.oss.OssPath
+import cromwell.filesystems.oss._
 
 trait GcsBatchCommandBuilder extends DefaultIoCommandBuilder {
```
Comment (danbills) [inline filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala]: I believe the changes int his file are unnecessary but we've asked @Horneth to chime in

File: engine/src/main/scala/cromwell/engine/io/IoActor.scala [hunk 13]
```diff
@@ -57,19 +61,24 @@ final class IoActor(queueSize: Int,
     val input = builder.add(Flow[IoCommandContext[_]])
     
     // Partitions requests between gcs batch, and single nio requests
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/io/IoActor.scala]: This comment is no longer accurate

File: filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala (hunk 12 above)
Comment (geoffjentry) [reply filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala]: Yeah, @crisish can you explain why you're piggybacking on GCS here?

**Comment by geoffjentry** [comment]
@crisish general comment - can you add `final` to all of your case classes? i know that in some cases you were mimicking other case classes which didn't have it but I'm hoping to not make that issue worse.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+  override def mapCommandLineWomFile(womFile: WomFile): WomFile = {
+    getPath(womFile.valueString) match {
+      case Success(ossPath: OssPath) =>
+        WomFile(WomSingleFileType, localizeOssPath(ossPath))
+      case _ => womFile
+    }
+  }
+
+
+  override lazy val commandLineValueMapper: WomValue => WomValue = {
```
Comment (geoffjentry) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: @crisish Can you explain more why the override is necessary? I know you put in some WDL above but I'm not familiar enough w/ OSS to know what's going.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsClusterIdOrConfiguration.scala [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+  val spotPattern = s"""$resourceAndInstanceAndImagePattern\\s+$spotStrategyPattern\\s+$spotPriceLimitPattern""".r
+
+
+  def parse(cluster: String): Try[BcsClusterIdOrConfiguration] = {
+    cluster match {
+      case idPattern(clusterId) => Success(Left(clusterId))
+      case instanceAndImagePattern(instanceType, imageId) => Success(Right(AutoClusterConfiguration(defaultResourceType, instanceType, imageId)))
+      case resourceAndInstanceAndImagePattern(resourceType, instanceType, imageId) => Success(Right(AutoClusterConfiguration(resourceType, instanceType, imageId)))
+      case spotPattern(resourceType, instanceType, imageId, spotStrategy, spotPriceLimit) => Success(Right(AutoClusterConfiguration(resourceType, instanceType, imageId, Some(spotStrategy), Some(spotPriceLimit.toFloat))))
+      case _ => Failure(new IllegalArgumentException("must be some string kike 'cls-xxxx' or 'OnDemand ecs.s1.large img-ubuntu'"))
```
Comment (geoffjentry) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsClusterIdOrConfiguration.scala]: kike == like

File: project/Dependencies.scala [hunk 16]
```diff
@@ -206,6 +206,19 @@ object Dependencies {
...
+      exclude("commons-beanutils", "commons-beanutils-core")
+      exclude("commons-collections", "commons-collections")
+  )
+
+  private val aliyunBatchcomputeDependencies = List(
```
Comment (mcovarr) [inline project/Dependencies.scala]: capitalize `Compute`

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+
+    callInputFiles ++ writeFunctionInputs
+  }
+
+  private[bcs] def generateBcsInputs(jobDescriptor: BackendJobDescriptor): Unit = {
+    getInputFiles(jobDescriptor) flatMap {
+      case (name, files) => bcsInputsFromWomFiles(name, files, jobDescriptor)
+    }
+
+    return
```
Comment (mcovarr) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: unnecessary

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJobPaths.scala [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+import cromwell.core.path.{DefaultPathBuilder, Path}
+
+object BcsJobPaths {
+    val BcsLogPathKey = "bcsLog"
+    val BcsEnvExecKey = "exec"
+	  val BcsEnvCwdKey = "cwd"
+	  val BcsEnvStdoutKey = "stdout"
+	  val BcsEnvStderrKey = "stderr"
+	  val BcsCommandDirectory: Path = DefaultPathBuilder.get("/cromwell_root")
+		val BcsTempInputDirectory: Path = DefaultPathBuilder.get("/cromwell_inputs")
```
Comment (mcovarr) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJobPaths.scala]: whitespace issues

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala (hunk 14 above)
Comment (kshakir) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: That example WDL above is an interesting use case. I don't know enough about OSS nor the new BCS backend at the moment, but I would never tell a WDL author to pass a path that way.

In ~`JES`~ `PAPI` (nor in the `Local` backend really) output paths are (I believe?) assumed to be relative. If a user tries to specify an absolute path like this they're going to have a bad time.

This WDL wouldn't be portable, and may not be 100% supported/recommended as it might try to overwrite the absolute path on a second run:
```wdl
workflow three_step {
  call ps {
    input: out = "gs://my-hard-coded-bucket/top/level/path/to/file"
  }
}
```

Instead I would tell the author to do:
```wdl
workflow three_step {
  call ps {
    input: out = "path/to/file"
  }
}
```

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala (hunk 8 above)
Comment (crisish) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala]: Thx, it works.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala (hunk 4 above)
Comment (crisish) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: Ok!

File: engine/src/main/scala/cromwell/engine/io/IoActor.scala (hunk 6 above)
Comment (crisish) [reply engine/src/main/scala/cromwell/engine/io/IoActor.scala]: Ok, these modifies are necessary before an oss nio interface is available, I forget to remove them. I'll solve this problem as soon as possible.

File: filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala (hunk 12 above)
Comment (crisish) [reply filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchCommandBuilder.scala]: I removed it.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala [hunk 19]
```diff
@@ -179,25 +178,36 @@ class BcsAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
...
       case Success(ossPath: OssPath) =>
         WomFile(WomSingleFileType, localizeOssPath(ossPath))
+      case Success(path: Path) if !path.isAbsolute =>
+        WomFile(WomSingleFileType, relativeOutputPath(path))
       case _ => womFile
     }
   }
 
 
+  /*
```
Comment (danbills) [inline supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsAsyncBackendJobExecutionActor.scala]: If unused, please delete.

**Comment by mcovarr** [comment]
There was a [bug](https://github.com/broadinstitute/cromwell/issues/3102) in the centaurCwlConformancePAPI test script that would have caused a false failure for you. The fixes for this bug are on develop so when you rebase this should no longer be an issue.

File: src/bin/travis/resources/local_centaur.conf [hunk 20]
```diff
@@ -65,8 +65,8 @@ backend.providers {
 
 database {
   db.url = "jdbc:mysql://localhost/cromwell_test?rewriteBatchedStatements=true"
-  db.user = "travis"
-  db.password = ""
+  db.user = "root"
```
Comment (geoffjentry) [inline src/bin/travis/resources/local_centaur.conf]: @crisish Why did you change this?

File: docs/backends/Backends.md [hunk 21]
```diff
@@ -54,6 +56,7 @@ The backend/filesystem pairings are as follows:
 
 * Local, HPC and Spark backend use the [Shared Local Filesystem](HPC/#filesystems).
 * Google backend uses the [Google Cloud Storage Filesystem](Google/#google-cloud-storage-filesystem).
+* Alibaba Cloud backend uses the OSS Storae FileSystem.
```
Comment (katevoss) [inline docs/backends/Backends.md]: > Storae

Should be "Storage"

File: docs/backends/BCS.md [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+
+```hocon
+backend {
+  providers {
+    default = BCS
+  }
+}
+```
+
+Before reading futher in this section please see the [Getting started on Alibaba Cloud](../tutorials/BCSIntro.md) for instructions of configuring to Alibaba Cloud services.
```
Comment (katevoss) [inline docs/backends/BCS.md]: > for instructions of configuring

I would change this to "for instructions **on** configuring"

File: docs/backends/BCS.md [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+        }
+      }
+    }
+  }
+}
+```
+
+### Runtime Attributes
+
+This backend support additional runtime attributes that are specified in the configuration key `backend.providers.BCS.config.runtime-attributes`.
```
Comment (katevoss) [inline docs/backends/BCS.md]: > This backend support additional

Should be "This backend support**s** additional"

File: docs/backends/BCS.md [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+      default-runtime-attributes {
+        cluster: "<resource-type> <instance-type> <image-id>"
+        # Maybe like cluster: "OnDemand ecs.sn1ne.large img-ubuntu"
+      }
+  ```
+
+#### mounts
+
+BCS jobs can mount an OSS object or an OSS prefix to local filesystem as a file or a directory in VM. 
+It uses distribute-caching and lazy-load techniques to optimize concurrently read reqeust of OSS file system.
```
Comment (katevoss) [inline docs/backends/BCS.md]: > It uses distribute-caching and lazy-load techniques to optimize concurrently read reqeust of OSS file system. 

"It uses distribute-caching and lazy-load techniques to optimize concurrently read **requests** of **the** OSS file system."

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssFileChannel.scala [hunk 25]
```diff
@@ -0,0 +1,19 @@
+package cromwell.filesystems.oss.nio
```
Comment (geoffjentry) [inline filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssFileChannel.scala]: @crisish Absolutely not a blocker for merging this but you all will want to consider splitting the nio implementation out into a separate library. That way GATK4 can make use of it as well

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala (hunk 8 above)
Comment (geoffjentry) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala]: @crisish Can you make sure to fold in the suggestion from @danbills ?

**Comment by danbills** [comment]
:+1:

**Comment by Horneth** [comment]
👍

File: project/Dependencies.scala [hunk 26]
```diff
@@ -246,6 +263,13 @@ object Dependencies {
     "com.readytalk" % "metrics3-statsd" % metrics3StatsdV
   )
 
+  val ossFileSystemDependencies = googleCloudDependencies ++ aliyunOssDependencies ++ List (
+    "com.github.pathikrit" %% "better-files" % betterFilesV
+  )
+
+  val gcsFileSystemDependencies = googleApiClientDependencies ++ googleCloudDependencies ++ List (
```
Comment (mcovarr) [inline project/Dependencies.scala]: This is not used anywhere

**Comment by geoffjentry** [comment]
@crisish The other thing which needs to be done is a rebase, it looks like there's a merge conflict

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssFileChannel.scala (hunk 25 above)
Comment (crisish) [reply filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssFileChannel.scala]: Ok, I'll try to update this NIO interface to our OSS team.

Comment (geoffjentry) [reply filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssFileChannel.scala]: @crisish Cool. Like I said, no need to do it before merging the PR, can always be done later

**Comment by crisish** [comment]
@geoffjentry Ok, I'm resolving this conflict.

File: supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala (hunk 8 above)
Comment (crisish) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala]: @geoffjentry It works but not perfect because I expect to return a None value if both `systemDisk` and `dataDisk` is empty.

Comment (geoffjentry) [reply supportedBackends/bcs/src/main/scala/cromwell/backend/impl/bcs/BcsJob.scala]: Ah ok. I was just going by the previous comment
